eventlet/hubs/selects.py

```diff
@@ -18,43 +18,62 @@ def _remove_bad_fds(self):
         """ Iterate through fds, removing the ones that are bad per the
         operating system.
         """
-        all_fds = list(self.listeners[READ]) + list(self.listeners[WRITE])
-        for fd in all_fds:
+        for fd in list(self.listeners_r) + list(self.listeners_w):
             try:
                 select.select([fd], [], [], 0)
             except select.error as e:
                 if get_errno(e) in BAD_SOCK:
                     self.remove_descriptor(fd)
 
     def wait(self, seconds=None):
-        readers = self.listeners[READ]
-        writers = self.listeners[WRITE]
-        if not readers and not writers:
-            if seconds:
-                time.sleep(seconds)
-            return
-        all_fds = list(readers) + list(writers)
-        try:
-            r, w, er = select.select(readers.keys(), writers.keys(), all_fds, seconds)
-        except select.error as e:
-            if get_errno(e) == errno.EINTR:
-                return
-            elif get_errno(e) in BAD_SOCK:
-                self._remove_bad_fds()
-                return
+
+        if not self.listeners_r and not self.listeners_w:
+            if seconds is not None:
+                ev_sleep(seconds)
+                if not self.listeners_r and not self.listeners_w:
+                    return
+                seconds = 0
             else:
-                raise
+                return
+
+        try:
+            r, w, er = select.select(list(self.listeners_r), list(self.listeners_w),
+                                     list(self.listeners_r)+list(self.listeners_w),
```

Comment:
`list(self.listeners_[rw])` is computed twice. This is a perfect example of low hanging fruit optimization.